Panic after calling Send (after x hours of functioning properly)

I'm not sure if this is a bug in the SDK (it looks similar to this issue: https://github.com/Azure/azure-service-bus-go/issues/141), or a implementation issue at my side that I'm not aware of, but after hours of functioning properly and publishing messages to the Azure Service Bus, my application panic-ed with the following trace:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb2110d]

goroutine 1072417 [running]:
github.com/Azure/azure-service-bus-go.(*Sender).Send(0xc00019f260, 0x1620840, 0xc0002b43c0, 0xc0008ab3b0, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/pkg/mod/github.com/!azure/azure-service-bus-go@v0.9.1/sender.go:151 +0x32d

github.com/Azure/azure-service-bus-go.(*Topic).Send(0xc0003978e0, 0x1620840, 0xc0002b43c0, 0xc0008ab3b0, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/pkg/mod/github.com/!azure/azure-service-bus-go@v0.9.1/topic.go:105 +0x1c1

github.com/brocaar/lora-app-server/internal/integration/azureservicebus.(*Integration).publish(0xc0002b4300, 0x1620900, 0xc000a9a900, 0x135d2cd, 0x2, 0x18, 0x270375d39373433, 0x131a280, 0xc0007c59a0, 0x0, ...)
/lora-app-server/internal/integration/azureservicebus/azure_service_bus.go:193 +0x6b9

github.com/brocaar/lora-app-server/internal/integration/azureservicebus.(*Integration).SendDataUp(0xc0002b4300, 0x1620900, 0xc000a9a900, 0x18, 0xc00098f6e0, 0x13, 0xc000534aa0, 0x10, 0x270375d39373433, 0xc000855a70, ...)
/lora-app-server/internal/integration/azureservicebus/azure_service_bus.go:78 +0xd7
```

As you can see in the trace, this is where the SDK gets called:
https://github.com/brocaar/lora-app-server/blob/v3.4.0/internal/integration/azureservicebus/azure_service_bus.go#L193

Before the panic, `Send` did publish a message to the Service Bus without error (e.g. there was no indication that there was a connection issue). This was in the logfiles before the panic:

```
integration/azureservicebus: event published" ctx_id=xxxxxxxx dev_eui=xxxxxxx event=up
```

(https://github.com/brocaar/lora-app-server/blob/v3.4.0/internal/integration/azureservicebus/azure_service_bus.go#L199)

Hi @devigned 

After upgrade from v0.7.0 to v0.10.0, we also has the same error.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8ee64f]

goroutine 11 [running]:
main/vendor/github.com/Azure/azure-service-bus-go.(*Sender).Send(0xc4201c47e0, 0xbfa7e0, 0xc4200220a0, 0xc4211ec780, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/azure-service-bus-go/sender.go:151 +0x34f
main/vendor/github.com/Azure/azure-service-bus-go.(*Queue).Send(0xc420370600, 0xbfa7e0, 0xc4200220a0, 0xc4211ec780, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/azure-service-bus-go/queue.go:167 +0x1a9
main/stream/asb.(*ASBStream).ProduceMessage(0xc420086820, 0xc42000ed48, 0x7f8f81a02601, 0xc420ca2120)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/stream/asb/asbs.go:185 +0xa7
main/stream/asb.(*ASBStream).ProduceMessageWithReconnect(0xc420086820, 0xc42000ed48, 0xb70522, 0x4)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/stream/asb/asbs.go:202 +0x4d
main/flow.(*T0ResultHandler).handle(0xc4200f9260, 0xbfe7e0, 0xc420d7e000, 0xc420b93260, 0x0, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/flow/result.go:101 +0x1f5
main/flow.(*VirtualAnalyzerResultHandler).handle(0xc4200f9280, 0xbfe7e0, 0xc420d7e000, 0xc420b93260, 0x0, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/flow/result.go:77 +0x22f
main/flow.(*ActionResultHandler).handle(0xc4200f92a0, 0xbfe7e0, 0xc420d7e000, 0xc420b93260, 0x0, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/flow/result.go:43 +0xcb
main/flow.WrapScanEventKafkaHandler.func1(0xc42006c5a0, 0xb9ed78, 0xbf4500, 0xc4200f92a0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/flow/scan.go:117 +0x82c
created by main/flow.WrapScanEventKafkaHandler
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/flow/scan.go:94 +0x17d
```

And, our consumer still has a panic when do reconnect.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x92f67a]

goroutine 95542 [running]:
main/vendor/github.com/Azure/go-amqp.(*Receiver).Receive(0x0, 0xeb4ec0, 0xc420406b40, 0xe14776, 0x1c, 0xeb4ec0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/go-amqp/client.go:1864 +0x3a
main/vendor/github.com/Azure/azure-service-bus-go.(*Receiver).listenForMessage(0xc4200cf600, 0xeb4ec0, 0xc420406b40, 0x0, 0x0, 0x0)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/azure-service-bus-go/receiver.go:338 +0x107
main/vendor/github.com/Azure/azure-service-bus-go.(*Receiver).listenForMessages(0xc4200cf600, 0xeb4ec0, 0xc420406b40, 0xc421831c80)
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/azure-service-bus-go/receiver.go:290 +0xef
created by main/vendor/github.com/Azure/azure-service-bus-go.(*Receiver).Listen
        /home/jenkinsbuild/ci-jenkins/workspace/TMCAS/cas-selfci-prod/prod/cas-gmail/go/src/main/vendor/github.com/Azure/azure-service-bus-go/receiver.go:212 +0x139
```

Any plan to fix this issue ?

Thanks

--- a/errors.go
+++ b/errors.go
@@ -35,6 +35,9 @@ type (
 	ErrNotFound struct {
 		EntityPath string
 	}
+
+	// ErrConnectionClosed indicates that the connection has been closed.
+	ErrConnectionClosed string
 )
 
 func (e ErrMissingField) Error() string {
@@ -79,4 +82,8 @@ func (e ErrNotFound) Error() string {
 func IsErrNotFound(err error) bool {
 	_, ok := err.(ErrNotFound)
 	return ok
+}
+
+func (e ErrConnectionClosed) Error() string {
+	return fmt.Sprintf("the connection has been closed: %s", string(e))
 }

--- a/sender.go
+++ b/sender.go
@@ -160,14 +160,8 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 		s.clientMu.RLock()
 		if s.session == nil {
 			// another goroutine has closed the connection
-			name := "Sender"
-			if s.Name != "" {
-				name = s.Name
-			}
 			s.clientMu.RUnlock()
-			err := ErrConnectionClosed(name)
-			tab.For(ctx).Error(err)
-			return err
+			return s.connClosedError(ctx)
 		}
 		msg.SessionID = &s.session.SessionID
 		next := s.session.getNext()
@@ -224,7 +218,14 @@ func (s *Sender) trySend(ctx context.Context, evt eventer) error {
 			return ctx.Err()
 		default:
 			// try as long as the context is not dead
+			s.clientMu.RLock()
+			if s.sender == nil {
+				// another goroutine has closed the connection
+				s.clientMu.RUnlock()
+				return s.connClosedError(ctx)
+			}
 			err = s.sender.Send(ctx, msg)
+			s.clientMu.RUnlock()
 			if err == nil {
 				// successful send
 				return err
@@ -264,6 +265,16 @@ func (s *Sender) trySend(ctx context.Context, evt eventer) error {
 			}
 		}
 	}
+}
+
+func (s *Sender) connClosedError(ctx context.Context) error {
+	name := "Sender"
+	if s.Name != "" {
+		name = s.Name
+	}
+	err := ErrConnectionClosed(name)
+	tab.For(ctx).Error(err)
+	return err
 }
 
 func (s *Sender) String() string {

--- a/sender.go
+++ b/sender.go
@@ -77,6 +77,7 @@ func (ns *Namespace) NewSender(ctx context.Context, entityPath string, opts ...S
 		}
 	}
 
+	// no need to take the write lock as we're creating a new Sender
 	err := s.newSessionAndLink(ctx)
 	if err != nil {
 		tab.For(ctx).Error(err)
@@ -96,7 +97,10 @@ func (s *Sender) Recover(ctx context.Context) error {
 	closeCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
 	closeCtx = tab.NewContext(closeCtx, span)
 	defer cancel()
-	_ = s.Close(ctx)
+	// we must close then rebuild the session/link atomically
+	s.clientMu.Lock()
+	defer s.clientMu.Unlock()
+	_ = s.close(ctx)
 	return s.newSessionAndLink(ctx)
 }
 
@@ -108,6 +112,11 @@ func (s *Sender) Close(ctx context.Context) error {
 	s.clientMu.Lock()
 	defer s.clientMu.Unlock()
 
+	return s.close(ctx)
+}
+
+// closes the connection.  callers *must* hold the client write lock before calling!
+func (s *Sender) close(ctx context.Context) error {
 	if s.doneRefreshingAuth != nil {
 		s.doneRefreshingAuth()
 	}
@@ -149,6 +158,17 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 
 	if msg.SessionID == nil {
 		s.clientMu.RLock()
+		if s.session == nil {
+			// another goroutine has closed the connection
+			name := "Sender"
+			if s.Name != "" {
+				name = s.Name
+			}
+			s.clientMu.RUnlock()
+			err := ErrConnectionClosed(name)
+			tab.For(ctx).Error(err)
+			return err
+		}
 		msg.SessionID = &s.session.SessionID
 		next := s.session.getNext()
 		s.clientMu.RUnlock()
@@ -259,12 +279,10 @@ func (s *Sender) getFullIdentifier() string {
 }
 
 // newSessionAndLink will replace the existing session and link
+// NOTE: this does *not* take the write lock, callers must hold it as required!
 func (s *Sender) newSessionAndLink(ctx context.Context) error {
 	ctx, span := s.startProducerSpanFromContext(ctx, "sb.Sender.newSessionAndLink")
 	defer span.End()
-
-	s.clientMu.Lock()
-	defer s.clientMu.Unlock()
 
 	connection, err := s.namespace.newClient()
 	if err != nil {

--- a/sender.go
+++ b/sender.go
@@ -24,13 +24,13 @@ package servicebus
 
 import (
 	"context"
-	common "github.com/Azure/azure-amqp-common-go/v3"
 	"sync"
 	"time"
 
+	common "github.com/Azure/azure-amqp-common-go/v3"
 	"github.com/Azure/azure-amqp-common-go/v3/uuid"
-	"github.com/devigned/tab"
 	"github.com/Azure/go-amqp"
+	"github.com/devigned/tab"
 )
 
 type (
@@ -148,8 +148,10 @@ func (s *Sender) Send(ctx context.Context, msg *Message, opts ...SendOption) err
 	defer span.End()
 
 	if msg.SessionID == nil {
+		s.clientMu.RLock()
 		msg.SessionID = &s.session.SessionID
 		next := s.session.getNext()
+		s.clientMu.RUnlock()
 		msg.GroupSequence = &next
 	}
 

